feat(nodes/ui): refactor field types, workflows

psychedelicious opened the pull request:
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [x] Bug Fix
- [x] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission


## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

      
## Have you updated all relevant documentation?
- [x] Yes - at least, I have added many comments to what was previously unintelligible code
- [ ] No


## Description

This PR is a substantial refactor of the frontend workflows logic.

There are some minimal changes to python code to support the frontend changes.

### Python Changes

- Deprecate most `UIType` enum members, which are no longer needed. If a
  deprecated type is used, a warning will be logged and the type ignored.
- Deprecate the `default_factory` arg to `InputField()` and `OutputField()`. If
  this arg is used, the callback will be called during invocation registration,
  and set as the `default` value. A warning will be logged.
  - This doesn't change anything in the core application - in the cases where a
    callback could actually do something (like a random seed), we actually
    always provided a valid value directly, so the callbacks were never actually
    called during graph execution.
  - In the future, we can explore restoring this functionality.
- Organization/renaming of helper classes (e.g. `_InputField` ->
  `InputJSONSchemaExtra`)
- Handle edge cases for field validation during invocation registration
- Surface the module name of custom nodes to the frontend as "node_pack"
- Improved/added comments throughout

Changes outside `baseinvocation.py` are minor and support these changes.

### Frontend Changes

The main frontend change is a refactor of fields, workflows and their types.

This involved renaming and shuffling many files and so there are a ton of small
changes throughout the code with new names or import paths.

#### Fields

The biggest changes are to fields, particularly field types.

##### Field Types

Previously, field types were strings. For example, integers would be
`"integer"`, `"IntegerCollection"` or `"IntegerPolymorphic"`.

Adding a new field type was a major pain, and there was a lot of fiddly logic
related to edge validation.

Field types are now structured. Here's integer:

```ts
type T = {
  name: 'IntegerField'; // literal
  isCollection: boolean;
  isPolymorphic: boolean;
};
```

##### Stateful and Stateless Field Types

Fields are now conceptually categorized as stateful or stateless.

Stateful fields are things like integers, strings or images. They have state
(value) in the frontend, and UI components allowing users to change their value.

Stateless fields are things like `UNetField` or `ControlField`. The user cannot
directly change these fields - they are created and consumed only in python - so
they have no UI component.

The signature of a stateless field type is a wider version of a stateful field:

```ts
type T = {
  name: string;
  isCollection: boolean;
  isPolymorphic: boolean;
};
```

The handling of fields is now generalized such that node authors may freely
create pydantic models for their fields. Many "official" field types are
stateless, and all "custom" field types in community nodes are stateless.

Connection validation still works as it did previously.

#### zod Schemas and Types

All workflow-related schemas and types have been rewritten, see files in
`invokeai/frontend/web/src/features/nodes/types/`.

Almost everything is now a zod schema with inferred TS types (previously there
was a lot of pure TS). The OpenAPI schema types are still pure TS.

The naming conventions for these things have been revised and made consistent.

There are many small changes in components and hooks resulting from newly named
schemas, types and type-guards.

#### Other Changes

- Migration logic for workflows implemented (see
  `invokeai/frontend/web/src/features/nodes/util/workflow/migrations.ts`)
  - We are now on workflow schema version 2.0.0. v1.0.0's schemas are retained
    in `invokeai/frontend/web/src/features/nodes/types/v1/` and used during
    migration.
- Community nodes' node packs are displayed in the info icon tooltip for a node
- Improved workflow loading (includes migration, validation, and improved user
  feedback)
- Improved node updating logic & UI/UX
- If a node in loaded workflow is missing from the user's install, the node type
  and node pack name will be displayed
- Linear UI parameter schemas are cleaned up and deduped (was duplication w/
  some field schemas)
- OpenAPI schema parsing logic simplified, edge cases handled
- Added many strings to translation system
- Updated README.md
- Added WORKFLOWS_DESIGN_IMPLEMENTATION.md, a high-level review of workflows and
  their implementation on frontend
- A number of minor bug fixes / edge cases
- Improved/added comments throughout
- Fixed polymorphic fields not able to be added to workflow linear view

## Related Tickets & Documents

<!--
For pull requests that relate or close an issue, please include them
below. 

For example having the text: "closes #1234" would connect the current pull
request to issue 1234.  And when we merge the pull request, Github will
automatically close the issue.
-->

- Related Issue #5142 
- Supersedes and Closes #5113
- Closes #4720

## QA Instructions, Screenshots, Recordings

Sorry for the size of this. I don't think I could split up the frontend changes into separate PRs without breaking the app - I rewrote a substantial portion of workflow-related logic across the frontend, and a partial implementation would leave it nonfunctional.

The changes are scoped to the workflow editor. Workflows that work on 3.4.0post2 should work in this PR.

You should be able to add custom fields. I made a [small node pack](https://github.com/psychedelicious/test-arbitrary-fields) with custom fields to test.

Everything in the workflow editor should work just like it did before.

RyanJDick reviewed (approved):
I reviewed the backend code and skimmed the frontend code. Overall, looks good. I left a few minor comments, but nothing that I think needs to hold up the PR.

I did some mild testing by re-basing https://github.com/invoke-ai/InvokeAI/pull/5142 on this branch and re-building the tiled upscaling workflow. I didn't run into any problems, but, again, this was far from an exhaustive test.

psychedelicious commented on invokeai/app/invocations/baseinvocation.py:
```diff
@@ -228,28 +267,59 @@ def InputField(
+
+    if default_factory is not _Unset and default_factory is not None:
+        default = default_factory()
+        del default_factory
```
Was a remnant from an earlier iteration where `default_factory` was in a dict.

psychedelicious commented on invokeai/app/invocations/baseinvocation.py:
```diff
@@ -266,70 +336,34 @@ def InputField(
     return Field(
         **provided_args,
-        json_schema_extra=json_schema_extra_,
+        json_schema_extra=json_schema_extra_.model_dump(exclude_none=True),
```
Without this option, the JSON has `"property_name": "null"` for every property that isn't valued. I had a reason for doing this, something related to the generated types or OpenAPI schema being slightly off without it. It's possible it's no longer needed.

psychedelicious replied:
I'm going to leave this as-is, I can't remember exactly where this broke things but I think it was in a subtle way. Another vote for clientside tests.

psychedelicious commented on invokeai/app/invocations/baseinvocation.py:
```diff
@@ -499,21 +542,29 @@ class BaseInvocation(ABC, BaseModel):
+        """Gets the invocation's type, as provided by the `@invocation` decorator."""
+        return cls.model_fields["type"].default
+
     @classmethod
     def register_invocation(cls, invocation: BaseInvocation) -> None:
+        """Registers an invocation."""
         cls._invocation_classes.add(invocation)
```
Should be straightforward to address this (separately). IIRC this wasn't a conscious design decision - just how it ended up.

psychedelicious replied:
https://github.com/invoke-ai/InvokeAI/issues/5190

psychedelicious commented on invokeai/app/invocations/custom_nodes/init.py:
```diff
@@ -47,5 +50,5 @@
 
-
-logger.info(f"Loaded {loaded_count} modules from {Path(__file__).parent}")
+if loaded_count > 0:
+    logger.info(f"Loaded {loaded_count} node packs from {Path(__file__).parent}")
```
I'll fix this in a followup.

psychedelicious commented on invokeai/app/invocations/primitives.py:
```diff
@@ -62,12 +62,12 @@ def invoke(self, context: InvocationContext) -> BooleanOutput:
     """A collection of boolean primitive values"""
 
-    collection: list[bool] = InputField(default_factory=list, description="The collection of boolean values")
+    collection: list[bool] = InputField(default=[], description="The collection of boolean values")
```
Yep! I think `default_factory` is very powerful, but I want to remove it before workflows start to take off so we have freedom with implementing it. I'll make an issue describing the challenges with supporting it.

psychedelicious replied:
https://github.com/invoke-ai/InvokeAI/issues/5187

psychedelicious commented on invokeai/frontend/web/docs/WORKFLOWS_DESIGN_IMPLEMENTATION.md:
```diff
@@ -0,0 +1,350 @@
+
+We use the item type for field type name, adding `isCollection: true` to the field type.
+
+##### Polymorphic Types
+
+When a field is annotated as a union of a type and list of that type, the schema object has an `anyOf` property, which holds a list of valid types for the union.
+
+After verifying that the union has two members (a type and list of the same type), we use the type for field type name, adding `isPolymorphic: true` to the field type.
```
At some point, I think we could support general unions. Inputs could have an `acceptsTypes: FieldType[]` property instead of a single `type` property. 

On the frontend, the workflow migration logic I've implemented would support a change like this. On the backend, it shouldn't be too hard to update the edge validation logic.

Now is the best time to change the term "polymorphic" to something more accurate, but I'm not sure what. Maybe "hybrid"?

psychedelicious replied:
Renamed to `CollectionOrScalar`

psychedelicious commented on invokeai/frontend/web/src/features/parameters/types/parameterSchemas.ts:
```diff
@@ -1,522 +1,269 @@
+  zParameterHeight,
 ]);
+export type ParameterResolution = z.infer<typeof zParameterResolution>;
+export const iParameterResolution = (
```
Thanks, removed. I ported these schemas over without checking if they were actually used.

psychedelicious commented:
@RyanJDick @maryhipp thanks for your review.

I've addressed most feedback or made an issue to address it in the future. 

@RyanJDick There are a couple points I'd like to resolve before merging this - left those convos open.

psychedelicious commented on invokeai/app/invocations/baseinvocation.py:
```diff
@@ -720,21 +800,30 @@ def wrapper(cls: Type[TBaseInvocation]) -> Type[TBaseInvocation]:
+        cls.UIConfig.tags = tags
+        cls.UIConfig.category = category
+
+        # Grab the node pack's name from the module name, if it's a custom node
+        module_name = cls.__module__.split(".")[0]
+        if module_name.endswith(CUSTOM_NODE_PACK_SUFFIX):
+            cls.UIConfig.node_pack = module_name.split(CUSTOM_NODE_PACK_SUFFIX)[0]
+        else:
+            cls.UIConfig.node_pack = None
```
We could do something like this:
```py
is_custom = cls.__module__.rsplit(".", 1)[0] == 'invokeai.app.invocations'
```

If a custom node structured its modules correctly, it could impersonate a core node.

Probably the better way to do this is to have `class CustomInvocation(BaseInvocation)` and `class CoreInvocation(BaseInvocation)`, and make it hard for custom nodes to import `CoreInvocation`. This is predicated on splitting the invocation registration logic from `BaseInvocation`.

The goal here isn't necessarily to prevent nodes from doing something malicious - its more to be able to differentiate core nodes from custom nodes in the interest of providing a good user and developer experience.

RyanJDick replied:
I wasn't even thinking from a security perspective - just seemed a little fragile. Let's not bother with trying to solve this right now. It's easy to change in the future. And then we can take the time to fully consider how we want to handle core vs. community.
